Clarifying SecurityError in StorageManager.getDirectory

mpsuzuki opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

<!-- 📝 For large changes, check our contribution guide:
     https://developer.mozilla.org/en-US/docs/MDN/Community/Pull_requests -->

<!-- 🚨 Low-quality, spammy, or disruptive pull requests are moderated heavily.
     Severe or repeated violations can result in being banned from contributing.
     Make sure you've read and agree to the following:
     - Community Participation Guidelines: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines
     - MDN Content Enforcement Policy: https://developer.mozilla.org/en-US/docs/MDN/Community/Community_Participation_Guidelines -->

<!-- Please complete all sections, especially the Motivation field. If there’s no related issue, explaining why this change is needed helps reviewers understand the context and prioritize your request. Without it, your PR may be delayed or closed. -->

<!-- 👷‍♀️ After submitting, go to the 'Checks' tab of your PR for the build status -->
---

### Description

Add concrete examples of the error in OPFS getDirectory() method, at [StorageManager.getDirectory#Exceptions](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory)

### Motivation

Currently, the documentation states:
```
SecurityError DOMException
  Thrown if the user agent is not able to map the requested directory to the local OPFS.
```
I propose adding concrete scenarios that cause this exception, as the lack of detail may lead to confusion for developers implementing OPFS-based storage, like:
```
SecurityError DOMException
  Thrown if the user agent is not able to map the requested directory to the local OPFS, due to factors such as:
    * Storage or memory constraints that prevent OPFS allocation.
    * Security policies, including restrictions in non-secure contexts (e.g., non-HTTPS sessions).
    * Private browsing mode, where OPFS access is commonly restricted by browsers to prevent persistent data storage.
```

### Additional details

SQLite developers have specifically highlighted OPFS limitations in Incognito and Guest modes in their documentation: [Achtung: Restrictions in Incognito and Guest Browsing Modes](https://sqlite.org/wasm/doc/trunk/persistence.md#incognito).

In real-world tests, both Firefox and Safari threw a SecurityError when calling getDirectory() inside Private Browsing mode:

Firefox 128.11.0 error: "Security error when calling GetDirectory"

Safari 17.6 error: "UnknownError: The operation failed for an unknown transient reason (e.g. out of memory)"

While browsers intentionally obscure Private Browsing detection, a small note in MDN can help developers anticipate storage failures in such contexts without directly exposing detection mechanisms.

Would it be possible to add a small clarification about these known limitations to better support developers encountering this issue?

### Related issues and pull requests

mpsuzuki committed:
Add examples of OPFS getDirectory() error.

wbamberg commented on files/en-us/web/api/storagemanager/getdirectory/index.md:
```diff
@@ -27,7 +27,10 @@ A {{jsxref('Promise')}} that fulfills with a {{domxref("FileSystemDirectoryHandl
-  - : Thrown if the user agent is not able to map the requested directory to the local OPFS.
+  - : Thrown if the user agent is not able to map the requested directory to the local OPFS, due to factors such as:
+    - **Storage or memory constraints** that prevent OPFS allocation.
+    - **Security policies**, including restrictions in non-secure contexts (e.g., non-HTTPS sessions).
```
Are there any other security policies here other than HTTPS? Because I don't see how that can be applicable here: if the document isn't HTTPS, this API just isn't available. So how could it throw in this case?

wbamberg replied:
More to the point, it's also documented in this reference page:

<img width="736" alt="Screen Shot 2025-06-11 at 10 18 46 AM" src="https://github.com/user-attachments/assets/77189da1-c1b7-429f-83e9-d4a68f726ef2" />

>  If I remove this bullet, your concern is resolved?

This concern, yes, but my other questions stand.

wbamberg replied:
> BTW, accidentally I found that some browser(s) allows OPFS under a HTTP session with localhost.

That's part of the definition of a secure context, see https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure:

> Locally-delivered resources such as those with http://127.0.0.1 URLs, http://localhost and http://*.localhost URLs (e.g., http://dev.whatever.localhost/), and file:// URLs are also considered to have been delivered securely.

mpsuzuki replied:
Thanks! Thus, the situation "the normal & private browsing modes of MS Edge allows, but the private browsing mode of Firefox prohibits getDirectory()" is consistent with their behaviors in HTTPS session.

wbamberg commented on files/en-us/web/api/storagemanager/getdirectory/index.md:
```diff
@@ -27,7 +27,10 @@ A {{jsxref('Promise')}} that fulfills with a {{domxref("FileSystemDirectoryHandl
+  - : Thrown if the user agent is not able to map the requested directory to the local OPFS, due to factors such as:
+    - **Storage or memory constraints** that prevent OPFS allocation.
+    - **Security policies**, including restrictions in non-secure contexts (e.g., non-HTTPS sessions).
+    - **Private browsing mode**, where OPFS access is commonly restricted by browsers to prevent persistent data storage.
```
Does this mean that the API doesn't work in private browsing mode? Our documentation on storage quotas and persistence (https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#private_browsing) says that in PB mode data is not persisted after the session ends, but not that the API is itself disabled. Do you have a test case where we can try this out?

mpsuzuki replied:
My testcase was found during my experiments with SQLite3 WASM, https://gitlab.com/mpsuzuki/test-sqlite3-wasm. I would update this repository by this weekend, to add local https servers for test.

In my experiment, the Private Browsing mode on Google Chrome and Microsoft Edge allow the clients to write something to OPFS. On the other hand, the Private Browsing mode on Firefox and Apple Safari disallow the clients to write anything on OPFS.

wbamberg replied:
It looks to me like, yes, Firefox throws when you call this method in PB mode. But in Safari, it looks like `navigator.storage` is just unavailable:

<img width="162" alt="Screen Shot 2025-06-11 at 10 40 13 AM" src="https://github.com/user-attachments/assets/a76b8a40-df5e-4255-b7da-7b16a6cb16e3" />

So I'm not seeing this (from https://gitlab.com/mpsuzuki/test-sqlite3-wasm#known-limitations):

> For example, in the Private Browsing mode, ...Apple Safari returns "Unhandled Promise Rejection: UnknownError: The operation failed for an unknown transient reason (e.g. out of memory)".

mpsuzuki replied:
Just I've updated my testkit [test-sqlite3-wasm](https://gitlab.com/mpsuzuki/test-sqlite3-wasm), HTTPS & HTTP servers are added.

wbamberg reviewed (commented):
Thanks for your PR. I had a couple of questions about some of the cases you have listed.

Also, do you have any references from the specification about when these errors are expected to be raised?

mpsuzuki committed:
Drop a bullet about the security policy.

The requirement of the SSL session is
already documented at the beginning of
OPFS documentation. So no need to repeat.
Suggested by wbambarg@github.

mpsuzuki commented:
@wbamberg , please could you tell me about the tasks unresolved in my side?

wbamberg commented on files/en-us/web/api/storagemanager/getdirectory/index.md:
```diff
@@ -27,7 +27,9 @@ A {{jsxref('Promise')}} that fulfills with a {{domxref("FileSystemDirectoryHandl
 
 - `SecurityError` {{domxref("DOMException")}}
-  - : Thrown if the user agent is not able to map the requested directory to the local OPFS.
+  - : Thrown if the user agent is not able to map the requested directory to the local OPFS, due to factors such as:
+    - **Storage or memory constraints** that prevent OPFS allocation.
+    - **Private browsing mode**, where OPFS access is commonly restricted by browsers to prevent persistent data storage.
```
```suggestion
  - : Thrown if the browser is not able to map the requested directory to the local OPFS, for example due to storage or memory constraints. Also thrown in some browsers if `getDirectory()` is called in private browsing mode.
- `UnknownError` {{domxref("DOMException")}}
  - : Thrown in some browsers if `getDirectory()` is called in private browsing mode.
```

maybe?

mpsuzuki committed:
Update files/en-us/web/api/storagemanager/getdirectory/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

wbamberg reviewed (approved):
👍 thank you!

mpsuzuki commented:
Thank you!
